adoc/chapters/programming_interface.adoc

```diff
@@ -19401,481 +19401,694 @@ that all of the built-in functions fulfill the same requirements for both
 host and device.
 
 
-=== Description of the built-in types available for SYCL host and device
+=== Description of generic type names
 
-All of the OpenCL built-in types are available in the namespace
-[code]#sycl#. For the purposes of this document we use
-generic type names for describing sets of valid SYCL types. The
-generic type names themselves are not valid SYCL types, but they
-represent a set of valid types, as defined in
-<<table.gentypes>>. Each generic type within a section is
-comprised of a combination of scalar, SYCL [code]#vec# and/or [code]#marray# class
-specializations. The letters [code]#{n}# and [code]#{N}# define valid sizes for
-class specializations, where [code]#{n}# means 2,3,4,8,16 and [code]#{N}# means
-any positive value of size_t type. Note that any reference to the base type
-refers to the type of a scalar or the element type of a SYCL [code]#vec# or
-[code]#marray# specialization.
+The functions defined in sections <<sec:math-functions>> through
+<<sec::relational-functions>> are specified using a set of generic type names.
+These are not actual type names available to SYCL applications.  Instead, they
+are a shorthand used in this document to describe the set of overloads for
+these functions.  For example the [code]#fmax# function is specified as:
 
+[source]
+----
+genfloat fmax(genfloat x, genfloat y)
+----
+
+The generic type [code]#genfloat# is a stand-in for a set of a actual type
+names: [code]#float#, [code]#double#, etc.  To construct the actual overloads
+that are available, replace the generic type name with the name of each of
+these actual types:
+
+[source]
+----
+float fmax(float x, float y)
+double fmax(double x, double y)
+template<int N> vec<float,N> fmax(vec<float,N> x, vec<float,N> y)
```

Comment:
Is there a benefit to keep the `vec` variants as templates? Since the set of valid values for `{n}` is so limited we could just have overloads for each one.

Reply:
I agree that this is possible.  I did it this way because I think this is what we agreed on when we discussed internal issue 278.

One advantage to implementing the `vec` versions as many overloads is that it allows implicit conversion from `__swizzled_vec__`.  For example:

```'
vec<float, 3> v{1.0f, 2.0f, 3.0f};
fmax(v.swizzle<2, 0, 1>(), v.swizzle<0, 1, 2>());
```

The calls to `v.swizzle` above return the unnamed type `__swizzled_vec__` which is convertible to `vec<float, 3>`.  If we implement `fmax` as a template, I believe the code above won't compile because template resolution doesn't consider conversions.

If we implement `fmax` as separate overloads for all of the `{n}` possibilities, then I think the code above does compile because overload resolution does consider conversions.

There is no similar issue with `marray` because there is no type like `__swizzled_vec__` that is convertible to an `marray`.  In any event, we cannot do multiple overloads for `marray` because there aren't a fixed number of `{N}` values.

@illuhad how does hipSYCL deal with the `vec` variants of these math functions?  Is the `{n}` a template parameter or are there many different overloads for all the `{n}` possibilities?

Reply:
It seems like this would be visible to applications, though, so this could cause portability problems.  For example, if the spec says the synopsis is:

```
template<int N> fmax(vec<N> x, vec<N> y)
```

Then, application code might do something like this:

```
template<int N>
void foo() {
  sycl::vec<N> x;
  sycl::vec<N> y;
  fmax<N>(x, y);
}
```

This won't compile unless the implementation actually implements `fmax` with a template parameter of type `int`.

Reply:
The spec should just say 1 line to rule them all:
```c++
template <typename X, typename Y> fmax(X x, Y y) -> return_t<X, Y>;
```
and just explain the conversion rule implemented under the hood and when it is valid, in a modern intensive generic way instead of in an old extensive overloaded way,
Having a special template for `int N` like
```c++
template<int N> vec<float,N> fmax(vec<float,N> x, vec<float,N> y)
```
looks like syntactic noise because it allows to write the cryptic
```c++
fmax<3>(1.f, 42.f) // returns a vfloat3 obviously o_o
```
instead of just
```c++
fmax(vfloat3 { 1.f }, 42.f) // returns a vfloat3 more naturally
``````

Reply:
> Is there a benefit to keep the `vec` variants as templates? Since the set of valid values for `{n}` is so limited we could just have overloads for each one.

I did this in a8bcf6296d683a377dc613a7c5bb53a2e6c3b215.